activerecord/lib/active_record/migration.rb

```diff
@@ -1140,11 +1140,15 @@ def current_version # :nodoc:
     end
 
     def needs_migration? # :nodoc:
-      pending_migration_versions.size > 0
+      !pending_migration_versions.empty?
     end
 
     def pending_migration_versions # :nodoc:
-      migrations.collect(&:version) - get_all_versions
+      migrations.map(&:version) - get_all_versions
+    end
+
+    def pending_migrations # :nodoc:
+      migrations.select { |migration| pending_migration_versions.include?(migration.version) }
```

Comment:
to get the full migration struct including scope (we do need it)

Or we can raise for multi dbs if VERSION or SCOPE passed, don't know what's better

I tried to allow VERSION with up direction here... but down direction won't work for some dummy versions like `0` (few tests failed with that)
So I guess we can support only UP direction here, but maybe raise for VERSION passed is ok as well

Reply:
@eileencodes this is the most interesting part, what do you think ?

Tests are green now, I didn't add new ones, but will do once we choose a strategy here

Reply:
we can also use `Set`(O(1) checks) here but I don't think if this optimization is needed, can migrations number be that high ?